##set working directory
setwd("/home/velazqam/Documents/Projects/CodonOptimization/GFPs_iterator")

#Read 4.5K sequences
GFPs = read.table("GFP3.tsv", sep="\t", header=F,stringsAsFactors=F)

colnames(GFPs) =c("Sequence","GLO_Score","GLO_LowestScore","GLO_LowestWordScore","CAI_Ubiquitous","CAI_Germline","CAI_Neurons","CAI_Somatic","CAI_1","TargetPies_3MM","TargetPies_4MM","TargetPies_5MM")

Sseqs=GFPs[c(4404,3293,2266,20),"Sequence"]

###Run 3 times algorithm
A="ATGCCCAAAAAGAAACGGAAAGTTTCTAAGGGCGAAGAACTTTTCACAGGAGTGGTTCCTATCCTCGTCGAACTGGACGGAGATGTGAACGGACACAAGTTCAGCGTGAGTGGGGAAGGTGAAGGAGATGCCACGTACGGCAAATTGACACTCAAGTTCATCTGCACCACCGGCAAACTTCCAGTTCCTTGGCCAACTCTTGTTACTACATTCTGCTATGGAGTTCAATGCTTTTCGCGCTACCCGGACCATATGAAAAGACACGACTTTTTTAAAAGTGCCATGCCAGAAGGATACGTGCAGGAACGAACCATCTTCTTCAAGGATGACGGAAACTATAAGACCCGTGCTGAAGTTAAGTTCGAAGGAGACACTCTTGTCAACCGAATCGAGTTAAAGGGAATCGATTTCAAGGAGGATGGCAACATTCTAGGTCATAAGCTCGAGTATAACTACAACAGTCATAACGTCTATATCATGGCTGACAAGCAGAAGAATGGCATTAAGGTCAACTTTAAGATTCGTCACAATATTGAAGACGGATCTGTCCAATTGGCAGATCACTACCAACAAAACACCCCAATTGGAGACGGACCAGTTCTGCTACCTGATAACCATTATTTGAGCACTCAATCTGCCCTCTCAAAGGATCCAAACGAAAAGCGTGACCACATGGTCCTCTTGGAATTCGTAACGGCAGCCGGAATCACCCATGGAATGGACGAATTGTACAAGTCCCGCCGTCGTAAGGCCAACCCAACAAAGCTTTCTGAAAACGCGAAGAAACTGGCCAAAGAGGTAGAGAACTAA"
B="ATGCCCAAAAAGAAACGGAAAGTTTCTAAGGGCGAAGAATTGTTCACTGGAGTTGTGCCTATCTTGGTTGAACTTGATGGAGATGTTAACGGTCATAAATTCTCTGTATCAGGAGAAGGAGAAGGCGACGCTACATACGGAAAGTTGACCTTGAAGTTCATCTGTACTACCGGAAAGCTTCCTGTTCCATGGCCAACGCTCGTGACCACATTCTGTTATGGCGTGCAGTGCTTTTCACGTTACCCGGACCATATGAAGAGACACGATTTCTTCAAGTCCGCTATGCCTGAGGGATATGTCCAAGAGCGAACCATCTTCTTTAAAGATGACGGAAACTATAAAACTAGAGCCGAGGTCAAGTTCGAGGGAGACACGCTTGTCAACAGAATCGAACTGAAGGGAATAGACTTCAAGGAAGATGGAAACATTCTTGGTCATAAGCTCGAGTACAACTACAACAGCCATAATGTTTATATCATGGCTGATAAGCAAAAGAACGGAATTAAAGTCAATTTTAAGATTCGTCACAACATCGAGGATGGATCAGTGCAGCTCGCTGATCACTACCAGCAAAACACGCCGATCGGTGATGGACCAGTCCTTTTGCCGGATAATCATTATCTCAGTACCCAGTCAGCACTTAGCAAGGATCCGAACGAAAAACGTGACCACATGGTTCTTCTCGAATTCGTGACCGCTGCTGGAATTACTCACGGAATGGATGAGTTGTACAAATCTAGACGTCGAAAGGCGAATCCAACAAAGCTCTCCGAGAACGCCAAGAAGCTTGCTAAGGAGGTCGAGAATTAA"
C="ATGCCCAAAAAGAAACGGAAAGTTTCTAAGGGCGAAGAATTGTTCACAGGCGTTGTCCCAATTTTAGTCGAGCTCGACGGTGATGTTAACGGACACAAGTTCTCCGTTTCGGGAGAGGGTGAGGGTGACGCTACGTATGGAAAGCTGACACTCAAGTTTATCTGTACTACTGGAAAGCTTCCGGTCCCATGGCCGACTTTGGTCACTACTTTTTGCTATGGAGTCCAATGCTTCTCTCGATACCCAGATCACATGAAGCGACACGATTTCTTTAAAAGTGCTATGCCAGAAGGATACGTTCAAGAACGCACCATTTTCTTTAAGGACGATGGCAACTATAAGACAAGAGCCGAAGTCAAATTCGAGGGAGACACCCTTGTAAATCGCATTGAACTCAAGGGCATCGACTTCAAGGAGGATGGAAATATCCTCGGACATAAACTTGAATATAATTATAACAGCCATAATGTGTACATCATGGCGGATAAACAGAAAAACGGTATCAAGGTTAATTTCAAGATTCGGCACAATATCGAAGACGGAAGCGTCCAGCTGGCGGACCACTACCAACAGAACACCCCAATCGGAGATGGACCAGTCTTGTTACCGGACAACCATTATCTCTCTACTCAAAGCGCCTTAAGCAAAGATCCAAATGAAAAGCGTGACCACATGGTCCTGCTCGAGTTCGTTACAGCTGCGGGCATTACTCATGGAATGGATGAGCTCTACAAATCACGTCGTCGCAAGGCTAACCCAACTAAGCTCTCAGAAAACGCAAAGAAACTCGCAAAAGAGGTCGAGAATTAA"

D1="ATGCCCAAAAAGAAACGGAAAGTTTCTAAGGGCGAAGAACTCTTTACCGGAGTCGTCCCAATTCTCGTCGAGCTCGACGGAGACGTAAACGGACACAAATTCTCGGTTTCCGGAGAAGGAGAAGGAGATGCTACTTATGGAAAACTCACCCTTAAATTCATTTGCACCACCGGAAAGCTCCCAGTGCCGTGGCCAACCCTTGTGACTACATTCTGCTACGGAGTTCAATGCTTTTCCCGCTACCCAGACCACATGAAGCGTCACGATTTCTTCAAATCAGCCATGCCAGAAGGATACGTCCAGGAGCGAACAATTTTCTTCAAGGACGATGGAAACTACAAGACTCGTGCTGAAGTCAAGTTTGAAGGAGATACTCTCGTGAACCGCATTGAGCTCAAGGGAATCGACTTCAAAGAAGATGGAAACATTCTTGGACACAAGCTCGAATACAACTATAACTCGCACAACGTGTATATCATGGCCGACAAGCAAAAAAATGGAATCAAGGTCAACTTCAAGATTCGCCACAACATCGAAGACGGGTCGGTTCAACTCGCTGATCACTACCAGCAGAACACACCAATTGGAGATGGACCAGTCCTCCTCCCTGATAATCACTACCTTTCCACTCAATCCGCTCTTAGCAAGGATCCAAATGAGAAAAGAGATCACATGGTTCTTCTCGAGTTTGTCACCGCCGCCGGAATCACCCACGGAATGGACGAGCTTTACAAGTCGAGACGCCGCAAGGCTAATCCAACCAAACTTTCTGAAAATGCCAAGAAACTTGCCAAGGAGGTCGAGAACTAA"
D2="ATGCCCAAAAAGAAACGGAAAGTTTCTAAGGGCGAAGAACTCTTTACCGGAGTCGTCCCAATTCTCGTCGAGCTCGACGGAGACGTAAACGGACACAAATTCTCGGTTTCCGGAGAAGGAGAAGGAGATGCTACTTATGGAAAACTCACCCTTAAATTCATTTGCACCACCGGAAAGCTCCCAGTGCCGTGGCCAACCCTTGTGACTACATTCTGCTACGGAGTTCAATGCTTTTCCCGCTACCCAGACCACATGAAGCGTCACGATTTCTTCAAATCCGCCATGCCAGAAGGATACGTGCAGGAGCGAACAATTTTCTTCAAGGACGATGGAAACTACAAGACCCGTGCTGAAGTCAAGTTTGAAGGAGATACTCTCGTGAACCGCATTGAGCTCAAGGGAATCGACTTCAAAGAAGATGGAAACATTCTTGGACACAAGCTCGAATACAACTATAACTCGCACAACGTGTATATCATGGCCGACAAGCAAAAAAATGGAATCAAGGTCAACTTCAAGATTCGCCACAACATCGAAGACGGGTCGGTTCAACTCGCTGATCACTACCAGCAGAACACACCAATTGGAGATGGACCAGTCCTCCTCCCTGATAATCACTACCTTTCCACTCAATCCGCTCTTAGCAAGGATCCAAATGAGAAAAGAGATCACATGGTTCTTCTCGAGTTTGTCACCGCCGCCGGAATCACCCACGGAATGGACGAGCTTTACAAGTCGAGACGCCGCAAGGCTAATCCAACCAAACTTTCTGAAAATGCCAAGAAACTTGCCAAGGAGGTCGAGAACTAA"

CA2="ATGCCCAAAAAGAAACGGAAAGTTTCTAAGGGCGAAGAACTCTTCACAGGAGTCGTCCCAATCCTCGTCGAGCTCGACGGTGACGTCAACGGACACAAGTTCTCCGTCTCAGGTGAGGGTGAGGGTGACGCCACCTACGGAAAGCTCACCCTCAAGTTCATCTGCACCACCGGAAAGCTCCCAGTCCCATGGCCAACCCTCGTCACCACCTTCTGCTACGGAGTCCAATGCTTCTCCCGTTACCCAGACCACATGAAGCGCCACGACTTCTTCAAGTCCGCCATGCCAGAGGGTTACGTGCAAGAGCGTACAATTTTTTTCAAGGACGATGGAAACTACAAGACCCGTGCCGAGGTCAAGTTCGAGGGAGACACCCTCGTCAACCGTATCGAGCTCAAAGGTATCGACTTCAAGGAGGACGGTAACATCCTCGGACACAAGCTCGAGTACAACTACAACTCCCACAACGTCTACATCATGGCCGACAAGCAAAAAAACGGAATCAAGGTCAACTTCAAGATCCGTCACAACATCGAGGACGGTTCCGTCCAACTCGCCGACCACTACCAACAGAATACCCCAATCGGAGACGGACCAGTCCTCCTCCCAGACAACCACTACCTCTCCACCCAATCCGCCCTCTCCAAGGACCCAAACGAGAAGCGTGACCACATGGTCCTCCTCGAGTTCGTCACCGCCGCCGGAATCACCCACGGTATGGACGAGCTCTACAAGTCCCGTCGTCGTAAGGCCAACCCAACCAAGCTCTCCGAGAACGCCAAGAAGCTCGCCAAGGAAGTCGAGAACTAA"

Sseqs=append(Sseqs, c(A, B, C, D1, D2, CA2))

##Now write to modify intron location

writeLines(Sseqs,"SelectedFPs2modify.txt")

intloc=c(82,247,399)

###Copy codons from:
Secu="ATGCCCAAAAAGAAACGGAAAGTTTCTAAGGGCGAAGAACTTTTTACTGGAGTTGTCCCAATTCTTGTCGAACTTGATGGAGATGTCAACGGACACAAATTCTCGGTTTCCGGAGAAGGAGAGGGTGATGCTACTTACGGAAAGCTTACCCTCAAATTTATTTGTACTACGGGTAAGCTCCCTGTACCATGGCCAACCCTCGTCACCACTTTCTGCTATGGAGTTCAATGTTTCTCTCGGTATCCAGATCACATGAAGAGACATGACTTTTTTAAATCGGCCATGCCAGAGGGATACGTACAGGAGAGGACTATCTTTTTTAAGGACGACGGAAACTACAAGACTCGAGCCGAAGTGAAGTTCGAGGGCGATACACTTGTTAACCGGATTGAGCTAAAGGGAATTGATTTCAAGGAGGATGGAAATATTCTGGGACACAAACTTGAATATAATTACAACTCGCACAATGTCTACATTATGGCTGATAAGCAAAAGAACGGAATTAAGGTCAATTTCAAGATTCGACATAACATTGAGGATGGAAGTGTCCAATTGGCTGATCATTATCAACAGAATACTCCAATAGGGGACGGTCCTGTTTTGTTGCCTGATAACCACTATTTGTCCACTCAAAGCGCTTTGTCAAAGGATCCAAACGAAAAACGCGATCATATGGTTCTCCTTGAGTTCGTAACGGCCGCTGGAATCACTCACGGTATGGATGAACTCTATAAATCTCGGCGGCGTAAGGCCAACCCTACTAAGTTGAGCGAGAATGCAAAAAAGCTTGCGAAGGAAGTGGAGAATTGA"

# +- 1aa closest to reading frame

stt=(intloc-(intloc %% 3)) - 2
edd=(intloc-(intloc %% 3)) + 3
bps=c()
for(i in 1:length(stt)){bps=append(bps,c(stt[i]:edd[i]))}


Secu=unlist(strsplit(Secu,""))

nSeqs=c()
for(seq in Sseqs){
seq=unlist(strsplit(seq,""))
seq[bps]=Secu[bps]
nSeqs=append(nSeqs,paste(seq,collapse="",sep=""))
}

##Redo tables
caiss <- data.frame(ca1 = rep(NA,length(nSeqs)),ca2 = rep(NA,length(nSeqs)),ca3 = rep(NA,length(nSeqs)),ca4 = rep(NA,length(nSeqs)),ca5 = rep(NA,length(nSeqs)))


for(i in 1:length(nSeqs)){
  caiss[i,]=c(CalculateCAI(nSeqs[i],CAIS,1,AAtoCodF),CalculateCAI(nSeqs[i],CAIS,2,AAtoCodF),CalculateCAI(nSeqs[i],CAIS,3,AAtoCodF),CalculateCAI(nSeqs[i],CAIS,4,AAtoCodF),CalculateCAI(nSeqs[i],CAIS,5,AAtoCodF)) 
}


nbpi <- data.frame(pi3 = rep(NA,length(nSeqs)),pi4 = rep(NA,length(nSeqs)),pi5 = rep(NA,length(nSeqs)))

for(i in 1:length(nSeqs)){
  nbpi[i,]=c(length(unique(Strfindpies(nSeqs[i],Pies,3))), length(unique(Strfindpies(nSeqs[i],Pies,4))), length(unique(Strfindpies(nSeqs[i],Pies,5)))) 
}

data=cbind(caiss,nbpi)

writeLines(nSeqs,"GFPs_sel.fasta")

write.table(data, "Values_sel.tsv",sep="\t", row.names=F, col.names=F)

### Read new values
SGFPs = read.table("GFP_sel.tsv", sep="\t", header=F,stringsAsFactors=F)

colnames(SGFPs) =c("Sequence","GLO_Score","GLO_LowestScore","GLO_LowestWordScore","CAI_Ubiquitous","CAI_Germline","CAI_Neurons","CAI_Somatic","CAI_1","TargetPies_3MM","TargetPies_4MM","TargetPies_5MM")

hengpi=c(1,1,11,6,3,1,5,3,0,0)
hengtosc=c(0,3,31.5,15,2,1.5,0,3,0,0)

###add data then introns then its done
uno="gtgagttttattatcgattttcgggattttctacttgtaattaccttttgtttttaataagttgcttttttccag"
dos="gtaggttttcagtttcaatttttgttgcaactacctcaactctatgttttcag"
tres="gtaaatttttcaagcttctaaattccaaaataaaccattttaatttcttaattccag"


ffgg=cbind(SGFPs,hengpi,hengtosc)
colnames(ffgg)[(ncol(ffgg)-1):ncol(ffgg)]=c("HengsPis","HengTotalScores")

nnggseqs=c()
for(seq in ffgg$Sequence){
  seq=unlist(strsplit(seq,""))
  nnggseqs=append(nnggseqs,paste(paste(seq[1:intloc[1]],sep="",collapse=""),uno,paste(seq[(intloc[1]+1):intloc[2]],sep="",collapse=""),dos,paste(seq[(intloc[2]+1):intloc[3]],sep="",collapse=""),tres,paste(seq[(intloc[3]+1):length(seq)],sep="",collapse=""),collapse="",sep=""))
}

ffgg$Sequence=nnggseqs

write.table(ffgg,"10_GFPs_vals.tsv",sep="\t",col.names=T,row.names=F, quote=F)


            
            
            
###Now read Germline expressed genes
ggenes=readLines("Top500_germline_lcapdev.seqs")

caissg <- data.frame(ca1 = rep(NA,length(ggenes)),ca2 = rep(NA,length(ggenes)),ca3 = rep(NA,length(ggenes)),ca4 = rep(NA,length(ggenes)),ca5 = rep(NA,length(ggenes)))


for(i in 1:length(ggenes)){
  caissg[i,]=c(CalculateCAI(ggenes[i],CAIS,1,AAtoCodF),CalculateCAI(ggenes[i],CAIS,2,AAtoCodF),CalculateCAI(ggenes[i],CAIS,3,AAtoCodF),CalculateCAI(ggenes[i],CAIS,4,AAtoCodF),CalculateCAI(ggenes[i],CAIS,5,AAtoCodF)) 
}


nbpig <- data.frame(pi3 = rep(NA,length(ggenes)),pi4 = rep(NA,length(ggenes)),pi5 = rep(NA,length(ggenes)))

for(i in 1:length(ggenes)){
  nbpig[i,]=c(length(unique(Strfindpies(ggenes[i],Pies,3))), length(unique(Strfindpies(ggenes[i],Pies,4))), length(unique(Strfindpies(ggenes[i],Pies,5)))) 
}

datag=cbind(caissg,nbpig)

write.table(datag, "Values_germlinegenes.tsv",sep="\t", row.names=F, col.names=F)

GG=read.table("GG_2addExp.tsv", sep="\t", header=F,stringsAsFactors=F)

load("minimal-data.RData")

exp=lcap.dt[as.character(GG[,1]),"Germline"]

GG=cbind(GG,exp)

###Correlation tests
#3 GLO score
#10 CAI
#11 pi 3
#12 pi 4
#13 pi 5
#14 expression data

cor.test(GG[,3],GG[,14], method="pearson")
cor.test(GG[,10],GG[,14], method="pearson")
cor.test(GG[,11],GG[,14], method="pearson")
cor.test(GG[,12],GG[,14], method="pearson")
cor.test(GG[,13],GG[,14], method="pearson")

library("ggpubr")

df= data.frame(Score=GG[,3],Expression=GG[,14])
ggscatter(df, x = "Score", y = "Expression", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "GLO score", ylab = "Germline Expression")

df= data.frame(Score=GG[,10],Expression=GG[,14])
ggscatter(df, x = "Score", y = "Expression", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Codon Adaptation Index", ylab = "Germline Expression")

df= data.frame(Score=GG[,11],Expression=GG[,14])
ggscatter(df, x = "Score", y = "Expression", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Targeting piRNAs (3MM allowed)", ylab = "Germline Expression")

df= data.frame(Score=GG[,12],Expression=GG[,14])
ggscatter(df, x = "Score", y = "Expression", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Targeting piRNAs (4MM allowed)", ylab = "Germline Expression")

df= data.frame(Score=GG[,13],Expression=GG[,14])
ggscatter(df, x = "Score", y = "Expression", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "Targeting piRNAs (5MM allowed)", ylab = "Germline Expression")

p = ggscatter(df, x = "Score", y = "Expression", 
              add = "reg.line", conf.int = TRUE, 
              cor.coef = TRUE, cor.method = "pearson",
              xlab = "Targeting piRNAs (5MM allowed)", ylab = "Germline Expression")

p + geom_point(data=df[18,], col="red") +  annotate(geom="text", x=df[18,1], y=df[18,2]-50, label=expression(paste(italic("znfx-1"))), color="red")


